cpg_pipes/targets.py

```diff
@@ -0,0 +1,640 @@
+Plus Sample properties: PedigreeInfo, Sex.
+"""
+import hashlib
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional
+import pandas as pd
+
+from . import Path
+from .types import AlignmentInput, CramPath, GvcfPath, SequencingType
+from .providers.storage import StorageProvider, Namespace
+
+logger = logging.getLogger(__file__)
+
+
+class Target:
+    """
+    Defines a target that stage can act upon. Classes like Sample, Dataset, Pipeline
+    extend this class.
+    """
+
+    def __init__(self):
+        # Whether to process even if outputs exist:
+        self.forced: bool = False
+        # If not set, exclude from the pipeline:
+        self.active: bool = True
+
+    def get_samples(self, only_active: bool = True) -> list['Sample']:
+        """
+        Get flat list of all samples corresponding to this target.
+        """
+        raise NotImplementedError
+
+    def get_sample_ids(self, only_active: bool = True) -> list[str]:
+        """
+        Get flat list of all sample IDs corresponding to this target.
+        """
+        return [s.id for s in self.get_samples(only_active=only_active)]
+
+    def alignment_inputs_hash(self) -> str:
+        """
+        Unique hash string of sample alignment inputs. Useful to decide
+        whether the analysis on the target needs to be rerun.
+        """
+        s = ' '.join(
+            sorted(
+                [
+                    str(s.alignment_input)
+                    for s in self.get_samples()
+                    if s.alignment_input is not None
+                ]
+            )
+        )
+        h = hashlib.sha256(s.encode()).hexdigest()[:38]
+        return f'{h}_{len(self.get_sample_ids())}'
+
+    @property
+    def target_id(self) -> str:
+        """
+        ID should be unique across target of all levels.
+
+        We are raising NotImplementedError instead of making it an abstract class, 
+        because mypy is not happy about binding TypeVar to abstract classes, see:
+        https://stackoverflow.com/questions/48349054/how-do-you-annotate-the-type-of
+        -an-abstract-class-with-mypy
+
+        Specifically,
+        ```
+        TypeVar('TargetT', bound=Target)
+        ```
+        Will raise:
+        ```
+        Only concrete class can be given where "Type[Target]" is expected
+        ```
+        """
+        raise NotImplementedError
+
+    def get_job_attrs(self) -> dict[str, str]:
+        """
+        Attributes for Hail Batch job.
+        """
+        raise NotImplementedError
+
+    def get_job_prefix(self) -> str:
+        """
+        Prefix job names.
+        """
+        raise NotImplementedError
+
+    def get_sequencing_type(self) -> SequencingType:
+        """
+        Sequencing type of samples. Will throw an error if more than
+        one sequencing type is found.
+        """
+        sequencing_types = set(s.sequencing_type for s in self.get_samples())
```

Comment:
As long as `SequencingType` is hashable, which I didn't think it was - maybe explicitly taking the string representation and then casting back to SequencingType.

Reply:
Enums are hashable, seems to be working good!

```
>>> set([SequencingType.GENOME, SequencingType.EXOME])
{<SequencingType.EXOME: 'exome'>, <SequencingType.GENOME: 'genome'>}
>>> set([SequencingType.GENOME, SequencingType.GENOME])
{<SequencingType.GENOME: 'genome'>}
```